org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/IndexBasedJavaSearchEnvironment.java

```diff
@@ -16,18 +16,23 @@
 import java.util.Iterator;
 import java.util.List;
 
+import org.eclipse.jdt.core.ICompilationUnit;
 import org.eclipse.jdt.core.IJavaProject;
 import org.eclipse.jdt.internal.compiler.env.INameEnvironment;
 import org.eclipse.jdt.internal.core.JavaProject;
 
 public class IndexBasedJavaSearchEnvironment {
 
-	public static INameEnvironment create(List<IJavaProject> javaProjects, org.eclipse.jdt.core.ICompilationUnit[] copies) {
-		Iterator<IJavaProject> next = javaProjects.iterator();
-		JavaSearchNameEnvironment result = new JavaSearchNameEnvironment(next.next(), copies);
+	public static INameEnvironment create(IJavaProject javaProject, ICompilationUnit[] copies) {
+		return new JavaSearchNameEnvironment(javaProject, copies);
+	}
+
+	public static INameEnvironment createWithReferencedProjects(IJavaProject javaProject, List<IJavaProject> referencedProjects, ICompilationUnit[] copies) {
+		JavaSearchNameEnvironment result = new JavaSearchNameEnvironment(javaProject, copies);
 
+		Iterator<IJavaProject> next = referencedProjects.iterator();
 		while (next.hasNext()) {
-			result.addProjectClassPath((JavaProject)next.next());
+			result.addProjectClassPath((JavaProject)next.next(), true);
```

Comment:
Should we only do this if we have mixed of projects like having at least on JDK9+ or JDK8- ? The reason for asking is, if we force exported only for all scenarios, will we loose some search results before a classpath is not exported in the settings?

And how will this effect the JDT.LS since in JDT.LS i'm not sure how the exported flag is used. @rgrunber ?

Reply:
If you look at the added test you'll ses the bug is reproduced with only Java 11 projects. Mixing Java 8 an Java 9+ was how the problem was found initially. But not the real cause.

I dont know about JDT.LS. If you have compile/module path consistently set up w.r.t. the search path, it should be fine. My understanding was, you dont use projects. I dont know how you specify classpaths, but best to do it consistently with JDT as used by Eclipse.